UTS-46 Mapping Tables

isomarcte opened the pull request:
UTS-46 mapping tables and associated codegen. Still needs some cleanup/tests/docs, but I wanted to get a draft PR for some early thoughts. 

FYI, I did start out using Scalameta to generate the code, but hit issues with the size of the generated AST (stack unsafe methods in the printer and huge amounts of memory required to generate the output), so I switched to a more stringy approach.

isomarcte commented:
@armanbilge the interesting code in is the local plugin. `CodePointMapping.scala` is all generated (and huge).

armanbilge commented:
This is exciting, taking a look now! Is your plan to keep the generated code committed, or should it eventually be a `src_managed` kind of thing?

isomarcte commented:
`UTS46IDNAMappingTable.scala` is documented, but I've not got to `LookupTableGen.scala` yet.

isomarcte commented:
> This is exciting, taking a look now! Is your plan to keep the generated code committed, or should it eventually be a src_managed kind of thing?

Not sure. Probably `src_managed`. I'm open to suggestion on that.

It's nice to be able to look at it without having to go into `target`, but having a big file in `src` that is generated annoying (also annoying for Scalacheck/Scalafmt). 

I think in Circe we have this type of thing in `src_managed`, but in http4s (I'm thinking MimeDb stuff) it's in src. Do you know why that is done differently?

armanbilge commented:
> in http4s (I'm thinking MimeDb stuff) it's in src. Do you know why that is done differently?

Historical reasons possibly. I'd be in favor of moving that scheme to `src_managed/` as well ... although we are revisiting the `MimeDB` thing before 1.0 anyway.

Other concerns are build reproducibility and possibly not needing the internet? Although the latter is a bit silly since dependencies and stuff have to come from the internet anyway.

armanbilge commented on project/LookupTableGen.scala:
```diff
@@ -0,0 +1,225 @@
+    A.toOrdering
+
+  def printAsHex(i: Int): String =
+    String.format("0x%04X", i.asInstanceOf[Object]) // Don't ask...
```
Would this work? 😅 
```suggestion
    String.format("0x%04X", new Integer(i)) // Don't ask...
```

isomarcte replied:
Apparently that's deprecated, but it does look like I can use `Integer.valueOf(i))`. I'll get that cleaned up.

armanbilge commented on project/LookupTableGen.scala:
```diff
@@ -0,0 +1,225 @@
+
+object LookupTableGen {
+
+  // Why is this not in cats?
+  private implicit def orderingFromOrder[A](implicit A: Order[A]): Ordering[A] =
+    A.toOrdering
```
Possibly creates a cycle if we have an implicit going the other direction? 🤔

isomarcte replied:
I suspect you're correct.

armanbilge commented:
This is really cool, thanks for getting the PR up!

What are the plans for testing? I believe there's also a file of tests available from unicode, which we can either generate sources from or just read dynamically at runtime (it will take some fiddling to do this on JS, but very doable).

Regarding the generated code: obviously it's all internal implementation details that we can iterate on and optimize in the future. But I was curious about how `Array` would compare to using `@switch`, or even some sort of trie-structure to try and "compress" the blocks of repeated values.

isomarcte commented:
For the `Array` idea, are you thinking we partition the code points into contiguous and non-contiguous `Array[Int]` values, then for contiguous ones just do `arr(i)` and for non-contiguous ones do a binary search? Or am I totally off base there?

isomarcte commented:
> What are the plans for testing? I believe there's also a file of tests available from unicode, which we can either generate sources from or just read dynamically at runtime (it will take some fiddling to do this on JS, but very doable).

Is there a test file for just the mapping stage of UTS-46? I didn't see one earlier, but perhaps I missed it.

My current plan was,

* For [0, 0x10FFFF], (all Unicode code points) ensure that we always get a sentinel value or a code point back from the mapping function.
* Look around for some testing data for UTS-46's mapping stage online
* If we can't find any/enough good testing data, then wait until we have the full UTS-46 algorithm in place and use the test data that Unicode provides. If it works for the full algorithm, it works for the mapping stage. That said, I'd like to have each stage tested as much as possible.

isomarcte commented on project/UTS46IDNAMappingTable.scala:
```diff
@@ -0,0 +1,920 @@
+/** A plugin for generating the UTS-46 lookup tables for step 1 of UTS-46
+  * processing.
+  *
+  * The plugin does not automatically run. It should only need to be run when
```
I'll need to update this if we move this to `src_managed`, as it will need to be always run.

armanbilge commented:
> For the `Array` idea, are you thinking we partition the code points into contiguous and non-contiguous `Array[Int]` values, then for contiguous ones just do `arr(i + baseOffset)` and for non-contiguous ones do a binary search? Or am I totally off base there?

Oh right, the non-contiguous blocks 🤔 A binary search seems reasonable, but of all the implementations I looked at I don't remember seeing that technique used (although I may have missed it). Tries seem to be a common choice.

---

> Is there a test file for just the mapping stage of UTS-46? I didn't see one earlier just the test data for the full UTS-46 algorithm, but perhaps I missed it.

Ah, right. I was thinking about this one which is the full algorithm.

https://www.unicode.org/Public/idna/14.0.0/IdnaTestV2.txt

Just spitballing: at least on the JVM, can we somehow test this against icu4j? I believe you can directly access the mapping table.

https://github.com/unicode-org/icu/blob/86166e0a2dd1119e086d04b8973dc486b4af8dcc/icu4j/main/classes/core/src/com/ibm/icu/impl/UTS46.java#L69

isomarcte commented:
@armanbilge I've rebased onto main. I moved the generated code to `src_managed` and squashed out the large file.

I'm increasingly skeptical of my stringy code gen stuff...I'm considering trying Scalameta again and micro batching the output to avoid large compile time heap usage and the stack unsafe printing issues...in any case I'm probably going to do a pretty large refactor/cleanup of the LookupTableGen. 

I'm also going to take a look at alternative methods for encoding the tables, e.g. tries etc. That may or may not be part of this PR. I want to get this generally stable and clean enough to merge so we can get a full IDNA implementation, then optimize.

In any case the non-generated code is getting closer to something stable, though still not all the way there yet. Let me know your thoughts on that when you get a chance.

armanbilge commented:
@isomarcte thanks for the update. For code-gen, have you considered tree hugger? That's what the MimeDB is using.

https://eed3si9n.com/treehugger/

> That may or may not be part of this PR. I want to get this generally stable and clean enough to merge so we can get a full IDNA implementation, then optimize.

I'm fully on-board with that :)

isomarcte commented:
@armanbilge I've not. I'll take a look at it.

armanbilge commented on core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala:
```diff
@@ -0,0 +1,220 @@
+    * as described in section 5 of UTS-46. useStd3ASCIIRules is true and
+    * transitionalProcessing is false.
+    */
+  def mapCodePoints(input: String): Either[MappingFailure, String] =
```
I wondered whether we could relax this to `CharSequence`, but it doesn't expose `codePointAt()`. I don't see why it couldn't to be honest.

Anyway, just a silly, unlikely optimization 😛

isomarcte replied:
Yeah, I've been kicking around changing it to `CharSequence`. The way the JVM's String/CharSequence APIs talk about characters, as distinct from (UTF-16) bytes, as distinct from Unicode code points, is all over the place and it makes it hard to abstract when you care both about UTF-16 bytes and Unicode code points.

What I really want is set typeclasses that allow you to inspect a StringLike values as either a sequence of bytes (in an arbitrary UTF) or distinctly as a sequence of Unicode Code points. And we could have instances for `String`, various collection types, anything which `<: CharSequence`, etc.

But, I'm trying not to get ahead of myself and that typeclass definitely shouldn't live here, though I don't know where it _should_ live.

It would be useful for idna4s, cats-uri, and case-insensitive, as well as just...in general.

isomarcte replied:
TL;DR I'm going to leave most things as `String` for now then do a pass at the end and see what can and should have a `CharSequence` or similar based API.

armanbilge commented on core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala:
```diff
@@ -0,0 +1,220 @@
+  def unsafeMapIntCodePoint(codePoint: Int): Either[String, CodePointStatus] =
+    CodePoint.fromInt(codePoint).map(mapCodePoint)
+
+  private val ReplacementCharacter: Int =
+    0xFFFD
```
This allows it to be inlined at use, both the `final` and lack of type annotation are necessary.

https://www.scala-lang.org/files/archive/spec/2.13/04-basic-declarations-and-definitions.html#value-declarations-and-definitions

```suggestion
  private final val ReplacementCharacter = 0xFFFD
```

isomarcte replied:
TIL. I'll update that.

armanbilge commented on core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala:
```diff
@@ -0,0 +1,220 @@
+import java.lang.StringBuilder
+
+private[uts46] trait CodePointMapperBase {
+  protected final val VALID: Int = -1
+  protected final val VALID_NV8: Int = -2
+  protected final val VALID_XV8: Int = -3
+  protected final val IGNORED: Int = -4
```
Same for all of these: I think if we can drop the `: Int` there's a good chance these values will be inlined. But since this is in a `trait` we'd have to check the bytecode to be sure 😅 

Also, can this be an `abstract class` instead?

armanbilge replied:
One more option is you could declare these in `CodePointMapper`, and then in `GeneratedCodePointMapper` do a self type `this: CodePointMapper.type =>`.

isomarcte replied:
Dropping the type ascription fixed the inlining. Thanks!

armanbilge commented on core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala:
```diff
@@ -0,0 +1,220 @@
+    import CodePointStatus._
+
+    (unsafeMapIntCodePointSentinel0(codePoint.value): @switch) match {
+      // Literals used rather than named constants because scalac won't
+      // generate a switch with the named constants.
+      case -1 => // VALID
```
Huh, that's annoying. I wonder if the `final val` in the `object` would make a difference, but I'm doubtful.

isomarcte replied:
I'll try `final val` and check.

armanbilge commented on core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala:
```diff
@@ -0,0 +1,220 @@
+      MappingErrorImpl(failureIndex, message, codePoint)
+  }
+
+  sealed abstract class MappingFailure extends RuntimeException with NoStackTrace {
```
`Exception`s are conventionally named something like `MappingFailureException` ... right?

isomarcte replied:
Sometimes...I'm not sure it is really a standard. See circe's `DecodingFailure` as a counter example. I don't mind changing it though.

armanbilge reviewed (commented):
I took a pass through the non-generated code. Overall it seems good, I just left a trail of nitpicks and micro-optimizations 😛 

Dumb question: we are publicly exposing a lot of `sealed` ADTs. How confident are we that supporting a new Unicode version or UTS46 revision or something wouldn't force us to break compatibility?

I.e. that they can make a change backwards-compatibly (they do maintain backwards-compatibility ... right?) that we would not be able to, because we promised our ADT was `sealed`.

I glanced at this but didn't see anything interesting.
https://unicode.org/versions/

isomarcte commented:
Update: I've a much better version of this on my branch https://github.com/isomarcte/idna4s/tree/tree-hugger. Unfortunately, tree-hugger's printer is also stack unsafe and I'm getting similar issues to what I got with Scalameta.

I'm considering a couple options.

isomarcte commented:
> Dumb question: we are publicly exposing a lot of sealed ADTs. How confident are we that supporting a new Unicode version or UTS46 revision or something wouldn't force us to break compatibility?

You mean `CodePointStatus` and `IDNA2008Status`? I'm very confident those will change between Unicode variants. 

From section 5 of UTS-46

> A table is provided for each version of Unicode starting with Unicode 5.1, in versioned directories under [[IDNA-Table](https://www.unicode.org/reports/tr46/#IDNATable)]. Each table for a version of the Unicode Standard will always be backward compatible with previous versions of the table: only characters with the status value disallowed may change in status or mapping value. Unlike the IDNA2008 table, this table is designed to be applied to the entire domain name, not just to individual labels. That design provides for the IDNA2003 handling of label separators. In particular, the table is constructed to forbid problematic characters such as U+2488 ( ⒈ ) DIGIT ONE FULL STOP, whose decompositions contain a "dot".

This doesn't explicitly say that they won't add a new status type (which would be the only way we'd break bincompat here)...but I doubt that would happen. If that did happen, then we'd have to create a new ADT for `CodePointStatus` to avoid binary breaking changes, but it should be noted that this type is only used in `  def mapCodePoint(codePoint: CodePoint): CodePointStatus` which isn't the primary interface. The actual IDNA algorithm will be operating directly on `String` values, so that will use `  def mapCodePoints(useStd3ASCIIRules: Boolean, transitionalProcessing: Boolean)(input: String): Either[MappingFailure, String]` which doesn't reference any of those types. Meaning even if we did have to deprecate that ADT, the impact of that deprecation would be minor. `mapCodePoint` is really nice for testing and understanding the system, but not actually needed for IDNA. I can add in an `Unknown` constructor to handle this though. Let me circle back to that.

isomarcte commented:
@armanbilge I still need to add tests and a few docs, but this is very close to done now.

I'm using `IntMap` and `BitSet` rather than the switches. ScalaMeta for code gen (I worked around the stack unsafe issues), and dropped a ton of code. 

I'd love your thoughts.

isomarcte commented:
The major issue with this approach is that the _first_ invocation of `mapCodePoints` or `mapCodePoint` is _very slow_. This is because the underlying static variables backing the various `IntMap` and `BitSet` values are not computed until the first invocation. That is, the `init` method on the class backing object `CodePointMapper` isn't invoked until then.

After the first invocation, things are _very fast_. 

I'm kicking around various ideas on how to improve this. Having the first invocation be slow here seems problematic to me, consider the first request on an http4s stack timing out due to this.

I can make initialization faster through various means, but most of them are trading compiled code size for initialization speed.

armanbilge commented:
I was wondering about that too. I looked at the generated code, are you using immutable collections to build them up? Might mutables be better here?

Also there's a bitset impl in cats collections with different performance characteristics (and support for sparsity I believe). Wort checking out.

I am also worried that on JS they need to become `lazy val`s to avoid becoming linked into JS if they are not actually used.

isomarcte commented:
> I was wondering about that too. I looked at the generated code, are you using immutable collections to build them up? Might mutables be better here?

I've tried various encodings using mutable builders, usually they are slightly slower than just using the `List.apply` literal.

armanbilge commented:
I was thinking specifically about the `IntMap` and `BitSet` actually, are those slower as well?

isomarcte commented:
Oh, `IntMap` is only in `immutable`. I tried building up with just a `mutable.HashMap` and converting the result to an `IntMap`, but that didn't help at all.

isomarcte commented:
I'll try with the `mutable.BitSet` (and take a look at the one in cats).

isomarcte commented:
@armanbilge hey, good news! Turns out most of the cost is in the `BitSet` construction so using `mutable.BitSet` yielded an 90% increase in initialization time for me.

On my system that is ~1.8 seconds. So still slow, but maybe fast enough? Not sure how we decide what fast enough is.

isomarcte commented:
@armanbilge There is still some tuning that could be done here, but do you think this is fast enough (in terms of initialization) for now? I'm planning a optimization step after all the main code is done.

armanbilge commented:
> So still slow, but maybe fast enough? Not sure how we decide what fast enough is.

Yeah tbh seems slow :( it depends on use-case. For a long-lived process annoying but whatever, but for a serverless Lambda or CLI or something you better be praying it never hits this code path 😂 

Also, this is effectively a long blocking process that will presumably eat up a thread on the Cate Effect compute pool. So not ideal 😅 

I'm afraid to measure what 1.8s translates to on JS.

---

With all that said, I think we should move forward with tests and such, and then get bogged down in these optimization nightmares. Worst case we can generate different code for different platforms maybe 🤔

isomarcte commented:
@armanbilge I took one more pass at it and I've got the init time to ~150ms. I don't really know how to benchmark init time, but my very basic benchmarks show that ~75ms is JVM startup before `init`.

I'll make a PR into this PR for my changes.

isomarcte commented:
@armanbilge over here: https://github.com/isomarcte/idna4s/pull/1

For some reason I can't add you a reviewer for PRs on my fork.

isomarcte commented:
@armanbilge I've moved on the documentation and testing. You'll see a few documentation and test commits trickle in over the next 1-2 days as I get time to wrap this PR up.

Thanks for all the reviews. You really helped make this a much better piece of code than my original iterations.

armanbilge commented:
Awesome!! Thanks for spear-heading these foundational projects and getting neck-deep in all the relevant standards :)

isomarcte commented:
@armanbilge ready for review

armanbilge commented:
Sorry I haven't gone through your latest changes, but I had a question about testing. Do you think my idea from https://github.com/typelevel/idna4s/pull/5#issuecomment-1235830921 could work here to directly test the mapping?

> Just spitballing: at least on the JVM, can we somehow test this against icu4j? I believe you can directly access the mapping table.
> 
> https://github.com/unicode-org/icu/blob/86166e0a2dd1119e086d04b8973dc486b4af8dcc/icu4j/main/classes/core/src/com/ibm/icu/impl/UTS46.java#L69

isomarcte commented:
That slipped my mind with all the codegen stuff. Let me take a look.

isomarcte commented:
@armanbilge test added using icu4j. https://github.com/typelevel/idna4s/pull/5/files#diff-a6cdc4c12abfc27eeea89d49658d28354a40f1f55d7ef462d4f5945d3dfdc09cR32

armanbilge commented on tests/jvm/src/test/scala/org/typelevel/idna4s/tests/uts46/CodePointMapperPlatformTests.scala:
```diff
@@ -0,0 +1,58 @@
+
+  // Note this test will become flaky idna4's and icu4j are targeting
+  // different versions of Unicode.
+  property("idna4s's uts-46 mapping step, should agree with icu4j's uts46-mapping step"){
```
Is it already flaky, or will it become flaky? I expect the icu4j version will have to be pinned.

isomarcte replied:
It may already be flaky. icu4j is using Unicode 14.0.0, [but as of September 13th 2022, we are using Unicode 15.0.0](https://www.unicode.org/versions/Unicode15.0.0/). 

icu4j has a `release-72-rc` branch that is on Unicode 15.0.0, but I don't see it on Maven central.

isomarcte replied:
Looks like there are changes in the mapping table in this revision. 

```diff
2,3c2,3
< # Date: 2021-07-10, 00:49:51 GMT
< # © 2021 Unicode®, Inc.
---
> # Date: 2022-05-02, 19:29:26 GMT
> # © 2022 Unicode®, Inc.
5c5
< # For terms of use, see http://www.unicode.org/terms_of_use.html
---
> # For terms of use, see https://www.unicode.org/terms_of_use.html
8c8
< # Version: 14.0.0
---
> # Version: 15.0.0
10c10
< # For documentation and usage, see http://www.unicode.org/reports/tr46
---
> # For documentation and usage, see https://www.unicode.org/reports/tr46
1281c1281,1282
< 0CF3..0CFF    ; disallowed                             # NA   <reserved-0CF3>..<reserved-0CFF>
---
> 0CF3          ; valid                                  # 15.0 KANNADA SIGN COMBINING ANUSVARA ABOVE RIGHT
> 0CF4..0CFF    ; disallowed                             # NA   <reserved-0CF4>..<reserved-0CFF>
1389c1390,1391
< 0ECE..0ECF    ; disallowed                             # NA   <reserved-0ECE>..<reserved-0ECF>
---
> 0ECE          ; valid                                  # 15.0 LAO YAMAKKAN
> 0ECF          ; disallowed                             # NA   <reserved-0ECF>
6209c6211,6212
< 10EB2..10EFF  ; disallowed                             # NA   <reserved-10EB2>..<reserved-10EFF>
---
> 10EB2..10EFC  ; disallowed                             # NA   <reserved-10EB2>..<reserved-10EFC>
> 10EFD..10EFF  ; valid                                  # 15.0 ARABIC SMALL LOW WORD SAKTA..ARABIC SMALL LOW WORD MADDA
6274c6277,6278
< 1123F..1127F  ; disallowed                             # NA   <reserved-1123F>..<reserved-1127F>
---
> 1123F..11241  ; valid                                  # 15.0 KHOJKI LETTER QA..KHOJKI VOWEL SIGN VOCALIC R
> 11242..1127F  ; disallowed                             # NA   <reserved-11242>..<reserved-1127F>
6446c6450,6452
< 11AF9..11BFF  ; disallowed                             # NA   <reserved-11AF9>..<reserved-11BFF>
---
> 11AF9..11AFF  ; disallowed                             # NA   <reserved-11AF9>..<reserved-11AFF>
> 11B00..11B09  ; valid                  ;      ; NV8    # 15.0 DEVANAGARI HEAD MARK..DEVANAGARI SIGN MINDU
> 11B0A..11BFF  ; disallowed                             # NA   <reserved-11B0A>..<reserved-11BFF>
6492c6498,6506
< 11EF9..11FAF  ; disallowed                             # NA   <reserved-11EF9>..<reserved-11FAF>
---
> 11EF9..11EFF  ; disallowed                             # NA   <reserved-11EF9>..<reserved-11EFF>
> 11F00..11F10  ; valid                                  # 15.0 KAWI SIGN CANDRABINDU..KAWI LETTER O
> 11F11         ; disallowed                             # NA   <reserved-11F11>
> 11F12..11F3A  ; valid                                  # 15.0 KAWI LETTER KA..KAWI VOWEL SIGN VOCALIC R
> 11F3B..11F3D  ; disallowed                             # NA   <reserved-11F3B>..<reserved-11F3D>
> 11F3E..11F42  ; valid                                  # 15.0 KAWI VOWEL SIGN E..KAWI CONJOINER
> 11F43..11F4F  ; valid                  ;      ; NV8    # 15.0 KAWI DANDA..KAWI PUNCTUATION CLOSING SPIRAL
> 11F50..11F59  ; valid                                  # 15.0 KAWI DIGIT ZERO..KAWI DIGIT NINE
> 11F5A..11FAF  ; disallowed                             # NA   <reserved-11F5A>..<reserved-11FAF>
6514c6528
< 1342F         ; disallowed                             # NA   <reserved-1342F>
---
> 1342F         ; valid                                  # 15.0 EGYPTIAN HIEROGLYPH V011D
6516c6530,6532
< 13439..143FF  ; disallowed                             # NA   <reserved-13439>..<reserved-143FF>
---
> 13439..1343F  ; disallowed                             # 15.0 EGYPTIAN HIEROGLYPH INSERT AT MIDDLE..EGYPTIAN HIEROGLYPH END WALLED ENCLOSURE
> 13440..13455  ; valid                                  # 15.0 EGYPTIAN HIEROGLYPH MIRROR HORIZONTALLY..EGYPTIAN HIEROGLYPH MODIFIER DAMAGED
> 13456..143FF  ; disallowed                             # NA   <reserved-13456>..<reserved-143FF>
6618c6634,6636
< 1B123..1B14F  ; disallowed                             # NA   <reserved-1B123>..<reserved-1B14F>
---
> 1B123..1B131  ; disallowed                             # NA   <reserved-1B123>..<reserved-1B131>
> 1B132         ; valid                                  # 15.0 HIRAGANA LETTER SMALL KO
> 1B133..1B14F  ; disallowed                             # NA   <reserved-1B133>..<reserved-1B14F>
6620c6638,6640
< 1B153..1B163  ; disallowed                             # NA   <reserved-1B153>..<reserved-1B163>
---
> 1B153..1B154  ; disallowed                             # NA   <reserved-1B153>..<reserved-1B154>
> 1B155         ; valid                                  # 15.0 KATAKANA LETTER SMALL KO
> 1B156..1B163  ; disallowed                             # NA   <reserved-1B156>..<reserved-1B163>
6671c6691,6693
< 1D246..1D2DF  ; disallowed                             # NA   <reserved-1D246>..<reserved-1D2DF>
---
> 1D246..1D2BF  ; disallowed                             # NA   <reserved-1D246>..<reserved-1D2BF>
> 1D2C0..1D2D3  ; valid                  ;      ; NV8    # 15.0 KAKTOVIK NUMERAL ZERO..KAKTOVIK NUMERAL NINETEEN
> 1D2D4..1D2DF  ; disallowed                             # NA   <reserved-1D2D4>..<reserved-1D2DF>
7704c7726,7728
< 1DF1F..1DFFF  ; disallowed                             # NA   <reserved-1DF1F>..<reserved-1DFFF>
---
> 1DF1F..1DF24  ; disallowed                             # NA   <reserved-1DF1F>..<reserved-1DF24>
> 1DF25..1DF2A  ; valid                                  # 15.0 LATIN SMALL LETTER D WITH MID-HEIGHT LEFT HOOK..LATIN SMALL LETTER T WITH MID-HEIGHT LEFT HOOK
> 1DF2B..1DFFF  ; disallowed                             # NA   <reserved-1DF2B>..<reserved-1DFFF>
7714c7738,7803
< 1E02B..1E0FF  ; disallowed                             # NA   <reserved-1E02B>..<reserved-1E0FF>
---
> 1E02B..1E02F  ; disallowed                             # NA   <reserved-1E02B>..<reserved-1E02F>
> 1E030         ; mapped                 ; 0430          # 15.0 MODIFIER LETTER CYRILLIC SMALL A
> 1E031         ; mapped                 ; 0431          # 15.0 MODIFIER LETTER CYRILLIC SMALL BE
> 1E032         ; mapped                 ; 0432          # 15.0 MODIFIER LETTER CYRILLIC SMALL VE
> 1E033         ; mapped                 ; 0433          # 15.0 MODIFIER LETTER CYRILLIC SMALL GHE
> 1E034         ; mapped                 ; 0434          # 15.0 MODIFIER LETTER CYRILLIC SMALL DE
> 1E035         ; mapped                 ; 0435          # 15.0 MODIFIER LETTER CYRILLIC SMALL IE
> 1E036         ; mapped                 ; 0436          # 15.0 MODIFIER LETTER CYRILLIC SMALL ZHE
> 1E037         ; mapped                 ; 0437          # 15.0 MODIFIER LETTER CYRILLIC SMALL ZE
> 1E038         ; mapped                 ; 0438          # 15.0 MODIFIER LETTER CYRILLIC SMALL I
> 1E039         ; mapped                 ; 043A          # 15.0 MODIFIER LETTER CYRILLIC SMALL KA
> 1E03A         ; mapped                 ; 043B          # 15.0 MODIFIER LETTER CYRILLIC SMALL EL
> 1E03B         ; mapped                 ; 043C          # 15.0 MODIFIER LETTER CYRILLIC SMALL EM
> 1E03C         ; mapped                 ; 043E          # 15.0 MODIFIER LETTER CYRILLIC SMALL O
> 1E03D         ; mapped                 ; 043F          # 15.0 MODIFIER LETTER CYRILLIC SMALL PE
> 1E03E         ; mapped                 ; 0440          # 15.0 MODIFIER LETTER CYRILLIC SMALL ER
> 1E03F         ; mapped                 ; 0441          # 15.0 MODIFIER LETTER CYRILLIC SMALL ES
> 1E040         ; mapped                 ; 0442          # 15.0 MODIFIER LETTER CYRILLIC SMALL TE
> 1E041         ; mapped                 ; 0443          # 15.0 MODIFIER LETTER CYRILLIC SMALL U
> 1E042         ; mapped                 ; 0444          # 15.0 MODIFIER LETTER CYRILLIC SMALL EF
> 1E043         ; mapped                 ; 0445          # 15.0 MODIFIER LETTER CYRILLIC SMALL HA
> 1E044         ; mapped                 ; 0446          # 15.0 MODIFIER LETTER CYRILLIC SMALL TSE
> 1E045         ; mapped                 ; 0447          # 15.0 MODIFIER LETTER CYRILLIC SMALL CHE
> 1E046         ; mapped                 ; 0448          # 15.0 MODIFIER LETTER CYRILLIC SMALL SHA
> 1E047         ; mapped                 ; 044B          # 15.0 MODIFIER LETTER CYRILLIC SMALL YERU
> 1E048         ; mapped                 ; 044D          # 15.0 MODIFIER LETTER CYRILLIC SMALL E
> 1E049         ; mapped                 ; 044E          # 15.0 MODIFIER LETTER CYRILLIC SMALL YU
> 1E04A         ; mapped                 ; A689          # 15.0 MODIFIER LETTER CYRILLIC SMALL DZZE
> 1E04B         ; mapped                 ; 04D9          # 15.0 MODIFIER LETTER CYRILLIC SMALL SCHWA
> 1E04C         ; mapped                 ; 0456          # 15.0 MODIFIER LETTER CYRILLIC SMALL BYELORUSSIAN-UKRAINIAN I
> 1E04D         ; mapped                 ; 0458          # 15.0 MODIFIER LETTER CYRILLIC SMALL JE
> 1E04E         ; mapped                 ; 04E9          # 15.0 MODIFIER LETTER CYRILLIC SMALL BARRED O
> 1E04F         ; mapped                 ; 04AF          # 15.0 MODIFIER LETTER CYRILLIC SMALL STRAIGHT U
> 1E050         ; mapped                 ; 04CF          # 15.0 MODIFIER LETTER CYRILLIC SMALL PALOCHKA
> 1E051         ; mapped                 ; 0430          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER A
> 1E052         ; mapped                 ; 0431          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER BE
> 1E053         ; mapped                 ; 0432          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER VE
> 1E054         ; mapped                 ; 0433          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER GHE
> 1E055         ; mapped                 ; 0434          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER DE
> 1E056         ; mapped                 ; 0435          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER IE
> 1E057         ; mapped                 ; 0436          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER ZHE
> 1E058         ; mapped                 ; 0437          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER ZE
> 1E059         ; mapped                 ; 0438          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER I
> 1E05A         ; mapped                 ; 043A          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER KA
> 1E05B         ; mapped                 ; 043B          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER EL
> 1E05C         ; mapped                 ; 043E          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER O
> 1E05D         ; mapped                 ; 043F          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER PE
> 1E05E         ; mapped                 ; 0441          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER ES
> 1E05F         ; mapped                 ; 0443          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER U
> 1E060         ; mapped                 ; 0444          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER EF
> 1E061         ; mapped                 ; 0445          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER HA
> 1E062         ; mapped                 ; 0446          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER TSE
> 1E063         ; mapped                 ; 0447          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER CHE
> 1E064         ; mapped                 ; 0448          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER SHA
> 1E065         ; mapped                 ; 044A          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER HARD SIGN
> 1E066         ; mapped                 ; 044B          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER YERU
> 1E067         ; mapped                 ; 0491          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER GHE WITH UPTURN
> 1E068         ; mapped                 ; 0456          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER BYELORUSSIAN-UKRAINIAN I
> 1E069         ; mapped                 ; 0455          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER DZE
> 1E06A         ; mapped                 ; 045F          # 15.0 CYRILLIC SUBSCRIPT SMALL LETTER DZHE
> 1E06B         ; mapped                 ; 04AB          # 15.0 MODIFIER LETTER CYRILLIC SMALL ES WITH DESCENDER
> 1E06C         ; mapped                 ; A651          # 15.0 MODIFIER LETTER CYRILLIC SMALL YERU WITH BACK YER
> 1E06D         ; mapped                 ; 04B1          # 15.0 MODIFIER LETTER CYRILLIC SMALL STRAIGHT U WITH STROKE
> 1E06E..1E08E  ; disallowed                             # NA   <reserved-1E06E>..<reserved-1E08E>
> 1E08F         ; valid                                  # 15.0 COMBINING CYRILLIC SMALL LETTER BYELORUSSIAN-UKRAINIAN I
> 1E090..1E0FF  ; disallowed                             # NA   <reserved-1E090>..<reserved-1E0FF>
7729c7818,7820
< 1E300..1E7DF  ; disallowed                             # NA   <reserved-1E300>..<reserved-1E7DF>
---
> 1E300..1E4CF  ; disallowed                             # NA   <reserved-1E300>..<reserved-1E4CF>
> 1E4D0..1E4F9  ; valid                                  # 15.0 NAG MUNDARI LETTER O..NAG MUNDARI DIGIT NINE
> 1E4FA..1E7DF  ; disallowed                             # NA   <reserved-1E4FA>..<reserved-1E7DF>
8216c8307,8308
< 1F6D8..1F6DC  ; disallowed                             # NA   <reserved-1F6D8>..<reserved-1F6DC>
---
> 1F6D8..1F6DB  ; disallowed                             # NA   <reserved-1F6D8>..<reserved-1F6DB>
> 1F6DC         ; valid                  ;      ; NV8    # 15.0 WIRELESS
8228c8320,8322
< 1F774..1F77F  ; disallowed                             # NA   <reserved-1F774>..<reserved-1F77F>
---
> 1F774..1F776  ; valid                  ;      ; NV8    # 15.0 LOT OF FORTUNE..LUNAR ECLIPSE
> 1F777..1F77A  ; disallowed                             # NA   <reserved-1F777>..<reserved-1F77A>
> 1F77B..1F77F  ; valid                  ;      ; NV8    # 15.0 HAUMEA..ORCUS
8231c8325,8326
< 1F7D9..1F7DF  ; disallowed                             # NA   <reserved-1F7D9>..<reserved-1F7DF>
---
> 1F7D9         ; valid                  ;      ; NV8    # 15.0 NINE POINTED WHITE STAR
> 1F7DA..1F7DF  ; disallowed                             # NA   <reserved-1F7DA>..<reserved-1F7DF>
8298c8393
< 1FA75..1FA77  ; disallowed                             # NA   <reserved-1FA75>..<reserved-1FA77>
---
> 1FA75..1FA77  ; valid                  ;      ; NV8    # 15.0 LIGHT BLUE HEART..PINK HEART
8304c8399,8400
< 1FA87..1FA8F  ; disallowed                             # NA   <reserved-1FA87>..<reserved-1FA8F>
---
> 1FA87..1FA88  ; valid                  ;      ; NV8    # 15.0 MARACAS..FLUTE
> 1FA89..1FA8F  ; disallowed                             # NA   <reserved-1FA89>..<reserved-1FA8F>
8308c8404
< 1FAAD..1FAAF  ; disallowed                             # NA   <reserved-1FAAD>..<reserved-1FAAF>
---
> 1FAAD..1FAAF  ; valid                  ;      ; NV8    # 15.0 FOLDING HAND FAN..KHANDA
8311c8407,8409
< 1FABB..1FABF  ; disallowed                             # NA   <reserved-1FABB>..<reserved-1FABF>
---
> 1FABB..1FABD  ; valid                  ;      ; NV8    # 15.0 HYACINTH..WING
> 1FABE         ; disallowed                             # NA   <reserved-1FABE>
> 1FABF         ; valid                  ;      ; NV8    # 15.0 GOOSE
8314c8412,8413
< 1FAC6..1FACF  ; disallowed                             # NA   <reserved-1FAC6>..<reserved-1FACF>
---
> 1FAC6..1FACD  ; disallowed                             # NA   <reserved-1FAC6>..<reserved-1FACD>
> 1FACE..1FACF  ; valid                  ;      ; NV8    # 15.0 MOOSE..DONKEY
8317c8416,8417
< 1FADA..1FADF  ; disallowed                             # NA   <reserved-1FADA>..<reserved-1FADF>
---
> 1FADA..1FADB  ; valid                  ;      ; NV8    # 15.0 GINGER ROOT..PEA POD
> 1FADC..1FADF  ; disallowed                             # NA   <reserved-1FADC>..<reserved-1FADF>
8319c8419,8420
< 1FAE8..1FAEF  ; disallowed                             # NA   <reserved-1FAE8>..<reserved-1FAEF>
---
> 1FAE8         ; valid                  ;      ; NV8    # 15.0 SHAKING FACE
> 1FAE9..1FAEF  ; disallowed                             # NA   <reserved-1FAE9>..<reserved-1FAEF>
8321c8422,8423
< 1FAF7..1FAFF  ; disallowed                             # NA   <reserved-1FAF7>..<reserved-1FAFF>
---
> 1FAF7..1FAF8  ; valid                  ;      ; NV8    # 15.0 LEFTWARDS PUSHING HAND..RIGHTWARDS PUSHING HAND
> 1FAF9..1FAFF  ; disallowed                             # NA   <reserved-1FAF9>..<reserved-1FAFF>
8344c8446,8447
< 2B739..2B73F  ; disallowed                             # NA   <reserved-2B739>..<reserved-2B73F>
---
> 2B739         ; valid                                  # 15.0 CJK UNIFIED IDEOGRAPH-2B739
> 2B73A..2B73F  ; disallowed                             # NA   <reserved-2B73A>..<reserved-2B73F>
8886c8989,8991
< 3134B..3FFFD  ; disallowed                             # NA   <reserved-3134B>..<reserved-3FFFD>
---
> 3134B..3134F  ; disallowed                             # NA   <reserved-3134B>..<reserved-3134F>
> 31350..323AF  ; valid                                  # 15.0 CJK UNIFIED IDEOGRAPH-31350..CJK UNIFIED IDEOGRAPH-323AF
> 323B0..3FFFD  ; disallowed                             # NA   <reserved-323B0>..<reserved-3FFFD>
```

isomarcte replied:
That is, some code points which our version of UTS-46 consider valid (because they are under Unicode 15.0.0), icu4j 71.0 considers disallowed (because they are under Unicode 14.0.0)

armanbilge replied:
Seems they have a tagged release as well:
https://github.com/unicode-org/icu/releases/tag/release-72-rc

Hopefully they'll have some published artifacts for that soon.

armanbilge commented on core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/IDNA2008Status.scala:
```diff
@@ -0,0 +1,65 @@
+
+  implicit val orderAndHashForIDNA2008Status: Order[IDNA2008Status] with Hash[IDNA2008Status] =
+    new Order[IDNA2008Status] with Hash[IDNA2008Status] {
+      override def hash(x: IDNA2008Status): Int =
+        1
```
Sorry if it's a dumb one: why is this hash a constant?

isomarcte replied:
I've no idea why I wrote that...

isomarcte replied:
which is to say, it should not be a constant.

isomarcte replied:
Also, that means I missed the law checks for this type. I'll add those in.

armanbilge commented on core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointStatus.scala:
```diff
@@ -0,0 +1,310 @@
+            1
+          // We don't need to handle Disallowed_STD3_Mapped and Unknown. One
+          // of the above cases will have matched Disallowed_STD3_Mapped and
+          // since there is currently now way to construct an Unknown, we
```
```suggestion
          // since there is currently no way to construct an Unknown, we
```

armanbilge commented on core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointStatus.scala:
```diff
@@ -0,0 +1,310 @@
+      of(NonEmptyList.one(mapping))
+  }
+
+  /**
+   * Special ADT constructor for binary compatibility.
+   *
+   * This type is present to allow greater latitude in case a binary breaking change is required
+   * in this ADT. It is not expected that it will ever be used.
+   */
+  sealed abstract class Unknown extends CodePointStatus
```
I understand the goal here about leaving the door open but I am not sure if this is the right way to achieve that. In the unlikely event we have to add another member to the ADT would it have to extend `Unknown`?

I think we either want to unseal `CodePointStatus` or make `Unknown` private. Either way users should be unable to prove that they've covered all the cases—which is what we want.

But honestly it's hard to reason about this. Would be interesting to add a compile-only test outside of the idna4s package that demonstrates that users are forced to consider a `case _ =>` for any total pattern match on `CodePointStatus`.

isomarcte replied:
So, after reviewing UTS-46 some more, I think we can just remove `Unknown`. They don't explicitly call out that they will never add new status types, but they do say that they will only ever move `disallowed` code points into the other status types.

I think that implies pretty strongly we don't need to worry about them adding new status types. If we do end up in that position, then we can create a new ADT and new functions, which would be very annoying, but given I'm pretty sure that won't happen, I'm okay with that risk.

Thoughts?

armanbilge replied:
👍 to removing.

armanbilge commented on core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala:
```diff
@@ -0,0 +1,449 @@
+     */
+    def errors: NonEmptyList[CodePointMappingException]
+
+    /**
+     * The input string, mapped as much as was possible. Code points which failed replaced with
+     * the Unicode replacement character � (0xFFFD). Returning this value on failure is mandated
+     * by UTS-46.
+     */
+    def partiallyMappedInput: String
```
Interesting that this is mandated. I pondered whether our return type should be shaped more like `Ior` or `(Option[MappingException], String)` but I think the `Either` is best understood.

isomarcte replied:
I don't want to use `Ior` because the `Ior.both` is a bit ambiguous on if the right side is really success or not. In this case, the `partiallyMappedInput` is definitely a failure.

armanbilge commented on core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala:
```diff
@@ -0,0 +1,449 @@
+        val value: Int = input.codePointAt(index)
+        val indexIncrement: Int = if (value >= 0x10000) 2 else 1
+
+        // We check valid then mapped first. This is for two reasons. First,
+        // we want to prioritize fast execution on the successful code
+        // path. Second, other than disallowed, valid/mapped make up the vast
+        // majority of the domain.
+
+        if (validAlways(value) || validNV8(value) || validXV8(value)) {
```
Considering the non-trivial cost of initializing the data structures, could we add one more short-circuit fast-path that just checks if the input is ascii? The goal being a user plugging nothing but ascii into here will never have to pay the price of initialization.

armanbilge replied:
Where by "ascii", I mean the valid sub-range of ascii or whatever 😛

isomarcte replied:
Yeah, that seems reasonable. I'll add that in.

armanbilge reviewed (commented):
Just a couple final comments!

isomarcte commented:
@armanbilge I also just added a commit to make the Unicode version we target explicit in the `build.sbt` and temporarily set it to `14.0.0` to avoid flakiness with the tests. I think before we release we should (explicitly) target the latest version of Unicode, but for now I want to avoid the flaky tests.

armanbilge commented:
I have no idea how it would be used downstream, but is it also worth `BuildInfo`ing the unicode version into our sources so it's available at runtime? I can do this in a follow-up PR if you think it could be useful.

isomarcte commented:
> I have no idea how it would be used downstream, but is it also worth BuildInfoing the unicode version into our sources so it's available at runtime? I can do this in a follow-up PR if you think it could be useful.

I have that here: https://github.com/isomarcte/idna4s/blob/uts-46-alt2/core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala#L448

I extract it out of the `IdnaMappingTable.txt` file during source code generation.

armanbilge commented on core/shared/src/main/scala/org/typelevel/idna4s/core/uts46/CodePointMapper.scala:
```diff
@@ -0,0 +1,449 @@
+   *   set. This includes the STD3 Valid and STD3 mapped variants. In other words, code points
+   *   which were disallowed by some prior Unicode version may be made allowed in some future
+   *   Unicode version or they may still be disallowed but have their mappings under STD3
+   *   changed or they may be considered ''valid under STD3 only'', when they were formerlly
+   *   disllowed under STD3.
```
Sorry I missed this before! But while we're here

```suggestion
   *   changed or they may be considered ''valid under STD3 only'', when they were formerly
   *   disallowed under STD3.
```

armanbilge commented:
@isomarcte:
- https://github.com/isomarcte/idna4s/pull/2

isomarcte commented:
@armanbilge I've merged your PR, added the ASCII fast path, and I believe I've addressed all your comments.

armanbilge reviewed (approved):
LGTM!! 🚀 thank you for this phenomenal effort!
